Implement consume_user_activation() web driver components

marcoscaceres opened the pull request:
Matches the [consume-user-activation](https://html.spec.whatwg.org/#user-activation-user-agent-automation) WebDriver command, which corresponds to these steps in HTML: https://html.spec.whatwg.org/#consume-user-activation

marcoscaceres commented on resources/testdriver.js:
```diff
@@ -959,6 +980,10 @@
+        },
+
+        async consume_user_activation () {
+            throw new Error("Consume_user_activation() is not implemented by testdriver-vendor.js");
```
```suggestion
            throw new Error("consume_user_activation() is not implemented by testdriver-vendor.js");
```

marcoscaceres commented on resources/testdriver.js:
```diff
@@ -1180,6 +1201,10 @@
-        }
+        },
+
+        async consume_user_activation () {
```
```suggestion
        async consume_user_activation() {
```

gsnedders reviewed (commented):
I think this is my only comment here?

gsnedders reviewed (commented):
To summarise what I just posted:

 * There's a `context` argument sometimes and not otherwise; it seems like you're currently just dropping it on the floor with the current implementation. I'd be happy to land this with no `context` argument (after all, that matches the WebDriver UX), which avoids needing to change much aside from the changing the JS side.
 * Adding the command to the `webdriver` module.

Also, can you also add some test to `/infrastructure/testdriver` to verify the `testdriver` command works where expected?

gsnedders commented on tools/wptrunner/wptrunner/executors/executorwebdriver.py:
```diff
@@ -425,6 +426,13 @@ def get_virtual_sensor_information(self, sensor_type):
+        self.webdriver = self.parent.webdriver
+
+    def consume_user_activation(self):
+        return self.webdriver.send_session_command("POST", "consume-user-activation")
```
Also, isn't the command `window/consume-user-activation`?

marcoscaceres replied:
Done in `87edc2ba5d` — now routes through `BrowserWindow.consume_user_activation()`.

marcoscaceres commented on tools/wptrunner/wptrunner/executors/executormarionette.py:
```diff
@@ -741,6 +742,12 @@ def remove_virtual_sensor(self, remove_parameters):
+    def setup(self):
+        self.marionette = self.parent.marionette
+
+    def consume_user_activation(self, sensor_type, sensor_params):
```
oops, copy pasta.

marcoscaceres commented on tools/wptrunner/wptrunner/executors/actions.py:
```diff
@@ -433,6 +433,16 @@ def __call__(self, payload):
+
+    def __call__(self, payload):
+        self.logger.debug("Consuming user activation")
+        return self.protocol.user_activation.consume_user_activation()
```
agree.

marcoscaceres replied:
Keeping this — `context` is handled centrally by `ActionContext` now, so the executor methods don't take it (same as `run_bounce_tracking_mitigations`).

marcoscaceres commented:
The `context` definitely matters, because you can deactivate cross origin frames... I saw a lot of other things take a `context` but also don't use it (so I has assumed magic happened... but it's also sad that people are dropping context on the floor).

Anyway, yes! Let's get that context passed around!!

gsnedders commented:
> The `context` definitely matters, because you can deactivate cross origin frames... I saw a lot of other things take a `context` but also don't use it on the python side (so I has assumed magic happened... but it's also sad that people are dropping context on the floor).
> 
> Anyway, yes! Let's get that context passed around!!

it's plausible there's some magic somewhere, and I'm just unaware of it! *digs around*

seemingly `ActionContext` in `tools/wptrunner/wptrunner/executors/base.py` does this, ultimately.

marcoscaceres committed:
Add test_driver.consume_user_activation() to testdriver.js

Adds the public `test_driver.consume_user_activation(context)` API mapping to the
HTML "consume user activation" WebDriver extension command, plus the
`test_driver_internal` not-implemented stub. The wptrunner/vendor implementations
that dispatch the actual command are added separately.

https://html.spec.whatwg.org/#user-activation-user-agent-automation

marcoscaceres committed:
Implement consume_user_activation() wptrunner executor components

Wires test_driver.consume_user_activation() through wptrunner to the
window/consume-user-activation WebDriver extension command: adds
ConsumeUserActivationProtocolPart, the action (plus its registry entry), and the
WebDriver executor implementation (Marionette raises NotImplementedError for now).
Adds the infrastructure/testdriver conformance test, and switches
html/user-activation/resources/utils.js off the Fullscreen proxy to the real command.

https://html.spec.whatwg.org/#user-activation-user-agent-automation

marcoscaceres committed:
Add testdriver docs entry and webdriver client method for consume_user_activation

marcoscaceres committed:
Add classic/consume_user_activation webdriver tests

marcoscaceres committed:
Fix consume_user_activation infra test: order-independent cases + expected-FAIL metadata where unimplemented

marcoscaceres committed:
Route consume_user_activation executor through the webdriver client method

marcoscaceres committed:
Return consume_user_activation() result directly in consumeTransientActivation

marcoscaceres commented on html/user-activation/resources/utils.js:
```diff
@@ -17,19 +17,12 @@ function getEvent(eventType) {
     return false;
-  }
+  await test_driver.consume_user_activation();
+  return !navigator.userActivation.isActive;
```
Fixed in `75ec150703` — now returns the command's boolean directly, so it no longer infers from `isActive` (removing the expiry race). On the manual-run point: the Fullscreen fallback was a pre-standardization workaround; now that this is a standardized WebDriver command, the helper relies on it like any other testdriver-based test (no sibling helper here keeps a fallback), so I've deliberately left it dropped.

marcoscaceres commented on infrastructure/metadata/infrastructure/testdriver/consume_user_activation.html.ini:
```diff
@@ -0,0 +1,6 @@
+[consume_user_activation.html]
+  [Nothing to consume]
+    expected: FAIL
+
+  [Consume transient activation]
+    expected: FAIL
```
Keeping the unconditional `FAIL` for now: unlike `set_permission` (which only fails on Safari), this command isn't implemented by any product yet, so it fails everywhere — unconditional is accurate today. Once an implementation lands, that product's XPASS will flag exactly where to scope it.

marcoscaceres committed:
Restore Fullscreen fallback in consumeTransientActivation for browsers without the command

whimboo commented on tools/webdriver/webdriver/client.py:
```diff
@@ -407,6 +407,9 @@ def minimize(self):
         return self.session.send_session_command("POST", "window/fullscreen")
 
+    def consume_user_activation(self):
+        return self.session.send_session_command("POST", "window/consume-user-activation")
```
To repeat here what I already wrote on https://github.com/whatwg/html/pull/8609#issuecomment-5024100443. Why is that command under `window`? Window is the OS level window which consume user activation is not related to. Instead it should be related to the current document's window and as such we should simply drop the `window/` prefix from the URL.
